File: milk_buckets.py
# USACO 2024 February Gold 

import sys

n = int(input())
cows = list( map(int, input().split()))

# 1. rotate the list so the smallest element is last
# 2. go backwards through the list, using binary search in insert into the list, making sure
#       That the last element in the list is the last item looped on
# 3. when going backwards, add the item & the shortest stack to another list
# 4. also make an index list

# 5. sum the milk at the start
# 6. go through each cow and its shortest stack, adding a delta milk (-x milk) at the correct index in a tracker list
# 7. go through the tracker list, adding each diff to the sum and printing

milk_sum = sum(cows) #5

def rotate(l, n):
    return l[n:] + l[:n]

class cow:
    def __init__ (self, m, i):
        self.m = m
        self.i = i

cows = rotate(cows, cows.index(min(cows)) + 1)

shortest = [cow(cows[n - 1], n - 1)]
dm = [0] * (n + 1)

for i in range(n - 2, -1, -1):
    cc = cow(cows[i], i)
    if cc.m > shortest[-1].m:
        shortest.append(cc)
    elif cc.m < shortest[-1].m:
        while cc.m < shortest[-1].m :
            shortest.pop()

        shortest.append(cc)
    else:
        shortest[-1].i = i

    current = cows[i]
    si =  i
    for z in range(len(shortest) - 1, -1, -1):
        dm[z.i - si] += shortest[z].m - current
        current = shortest[z].m

for e in range(1, n):
    milk_sum += dm[e]
    sys.stdout.write(str(milk_sum))
    sys.stdout.write('\n')

sys.stdout.write(str(milk_sum))

File: milk_buckets.java
import java.util.ArrayList;
import java.util.Scanner;

class Cow {
    int m, i;

    public Cow(int m, int i) {
        this.m = m;
        this.i = i;
    }
}

public class milk_buckets {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        int n = sc.nextInt();
        int[] cows = new int[n];
        for (int i = 0; i < n; i++) {
            cows[i] = sc.nextInt();
        }

        // 1. rotate the list so the smallest element is last
        cows = rotate(cows, findMinIndex(cows) + 1);

        // 5. sum the milk at the start
        int milkSum = sumArray(cows);

        ArrayList<Cow> shortest = new ArrayList<>();
        shortest.add(new Cow(cows[n - 1], n - 1));

        int[] dm = new int[n + 1];

        // 2. Go backwards through the list, similar to binary search logic
        for (int i = n - 2; i >= 0; i--) {
            Cow cc = new Cow(cows[i], i);

            if (cc.m > shortest.get(shortest.size() - 1).m) {
                shortest.add(cc);
            } else if (cc.m < shortest.get(shortest.size() - 1).m) {
                while (shortest.size() > 0 && cc.m < shortest.get(shortest.size() - 1).m) {
                    shortest.remove(shortest.size() - 1);
                }
                shortest.add(cc);
            } else {
                shortest.get(shortest.size() - 1).i = i;
            }

            // 6. Calculate delta milk at the correct index
            int current = cows[i];
            int si = i;
            for (int z = shortest.size() - 1; z >= 0; z--) {
                Cow zCow = shortest.get(z);
                dm[zCow.i - si] += zCow.m - current;
                current = zCow.m;
            }
        }

        // 7. Calculate and print the final milk sums
        for (int e = 1; e < n; e++) {
            milkSum += dm[e];
            System.out.println(milkSum);
        }

        System.out.println(milkSum);

        sc.close();
    }

    // 1. Rotate the array
    public static int[] rotate(int[] arr, int n) {
        int[] rotated = new int[arr.length];
        int index = 0;
        for (int i = n; i < arr.length; i++, index++) {
            rotated[index] = arr[i];
        }
        for (int i = 0; i < n; i++, index++) {
            rotated[index] = arr[i];
        }
        return rotated;
    }

    // Find index of the minimum element in the array
    public static int findMinIndex(int[] arr) {
        int minIndex = 0;
        for (int i = 1; i < arr.length; i++) {
            if (arr[i] < arr[minIndex]) {
                minIndex = i;
            }
        }
        return minIndex;
    }

    // Sum the elements of the array
    public static int sumArray(int[] arr) {
        int sum = 0;
        for (int value : arr) {
            sum += value;
        }
        return sum;
    }
}
